app/renderer/reducers/urlBarReducer.js

```diff
@@ -201,62 +200,73 @@ const generateNewSuggestionsList = (state) => {
     }
   }
 
-  const historyFilter = (site) => {
-    if (!site) {
-      return false
-    }
-    const title = site.get('title') || ''
-    const location = site.get('location') || ''
+  // NOTE: Iterating sites can take a long time! Please be mindful when
+  // working with the history and bookmark suggestion code.
+  const historySuggestionsOn = getSetting(settings.HISTORY_SUGGESTIONS)
+  const bookmarkSuggestionsOn = getSetting(settings.BOOKMARK_SUGGESTIONS)
+  const shouldIterateSites = historySuggestionsOn || bookmarkSuggestionsOn
+  if (shouldIterateSites) {
     // Note: Bookmark sites are now included in history. This will allow
     // sites to appear in the auto-complete regardless of their bookmark
     // status. If history is turned off, bookmarked sites will appear
     // in the bookmark section.
-    return (title.toLowerCase().includes(urlLocationLower) ||
-            location.toLowerCase().includes(urlLocationLower)) &&
-            site.get('lastAccessedTime')
-  }
-  var historySites = sites.filter(historyFilter)
+    const sitesFilter = (site) => {
+      const location = site.get('location')
+      if (!location) {
+        return false
+      }
+      const title = site.get('title')
+      return location.toLowerCase().includes(urlLocationLower) ||
+        (title && title.toLowerCase().includes(urlLocationLower))
+    }
 
-  // potentially append virtual history items (such as www.google.com when
-  // searches have been made but the root site has not been visited)
-  historySites = historySites.concat(suggestion.createVirtualHistoryItems(historySites))
+    let historySites = new Immutable.List()
+    let bookmarkSites = new Immutable.List()
+    const sites = appStoreRenderer.state.get('sites')
+    sites.forEach(site => {
+      if (!sitesFilter(site)) {
```

Comment:
Note for reviewers: this is most important change, to filter out non-matching sites as soon as possible to avoid downstream property lookup